### PR TITLE
Lucene-8920: refactor FST.Arc and utilities

### PR DESCRIPTION
**Pull request by msokolov**
This PR contains three commits; I recommend looking at them separately since at least the first one is pretty big and mostly noise. Overall there should be no functional change, just reorg and cleanup.

The first commit makes Arc (and some other FST) members private and adds accessors (setters and getters).

The second commit refactors code that uses the setters, and removes the setters so that Arcs can only be modified from within the FST class. 

The main change here was around the way that consumers would navigate from Arc to Arc by setting the internal state of an Arc to the desired index or position in the FST, and then call  `FST.readNextArc` to populate the remainder of the Arc members. With this change, Arc navigation is now done by calling either `FST.readArcByIndex` or `FST.readArcAtPosition`, which do the same thing that was being done before, just moving the code into `FST`.

One other change needed was in the memory codec. It was modifying Arc.output in order to accumulate its result. Instead it now initially copies the Arc.output and subsequently modifies its own private copy.

Finally, the third commit here refactors Arc's binarySearch implementation, which is repeated in several places into a method in `fst.Util`. I also added a few tests (in `TestUtil`) to explicitly test methods in `Util`, uncovering an off-by-one that had crept in to `Util.readCeilArc` during the refactor. That didn't seem to be caught by testing in the blocktreeords codec, its only user, but perhaps I just didn't run enough iterations.

I plan to follow up with subsequent commits for functional improvements; when I do I'll open a separate PR.

### CODE REVIEW

File: lucene/codecs/src/java/org/apache/lucene/codecs/memory/FSTOrdTermsReader.java [hunk 1]
```diff
@@ -305,7 +305,7 @@ public String toString() {
     }
 
     // Only wraps common operations for PBF interact
-    abstract class BaseTermsEnum extends org.apache.lucene.index.BaseTermsEnum {
+    abstract class  BaseTermsEnum extends org.apache.lucene.index.BaseTermsEnum {
```
[the remark this answers is not in the document]
Comment (msokolov) [reply lucene/codecs/src/java/org/apache/lucene/codecs/memory/FSTOrdTermsReader.java]: Yeah, not sure how that crept in. I'll fix this and merge shortly.

**Comment by msokolov** [comment]
pushed from my dev box
